controller/aks-cluster-config-handler.go

```diff
@@ -498,386 +479,132 @@ func BuildUpstreamClusterState(ctx context.Context, secretsCache wranglerv1.Secr
-		}
-
-		if spec.NetworkPolicy != nil {
-			networkProfile.NetworkPolicy = containersv211.NetworkPolicy(*spec.NetworkPolicy)
-		}
+	agentPoolClient, err := aks.NewAgentPoolClient(credentials)
+	if err != nil {
+		return spec, err
 	}
 
-	var agentPoolProfiles []containersv211.ManagedClusterAgentPoolProfile
-	if hasAgentPoolProfile(spec) {
+	needsUpdate := false
+	// check NodePools for update
+	if spec.NodePools != nil {
+		upstreamNodePools := utils.BuildNodePoolMap(upstreamSpec.NodePools)
 		for _, np := range spec.NodePools {
-			if np.OrchestratorVersion == nil {
-				np.OrchestratorVersion = spec.KubernetesVersion
-			}
-			agentProfile := containersv211.ManagedClusterAgentPoolProfile{
-				Name:                np.Name,
-				Count:               np.Count,
-				MaxPods:             np.MaxPods,
-				OsDiskSizeGB:        np.OsDiskSizeGB,
-				OsType:              containersv211.OSType(np.OsType),
-				VMSize:              containersv211.VMSizeTypes(np.VMSize),
-				Mode:                containersv211.AgentPoolMode(np.Mode),
-				OrchestratorVersion: np.OrchestratorVersion,
-			}
-			if np.AvailabilityZones != nil {
-				agentProfile.AvailabilityZones = np.AvailabilityZones
+			upstreamNodePool, ok := upstreamNodePools[to.String(np.Name)]
+			if ok {
+				// There is a matching node pool in the cluster already, so update it if needed
+				if to.Int32(np.Count) != to.Int32(upstreamNodePool.Count) {
+					logrus.Infof("Updating node count in node pool [%s] for cluster [%s]", to.String(np.Name), spec.ClusterName)
+					needsUpdate = true
+				}
+				if np.EnableAutoScaling != nil && to.Bool(np.EnableAutoScaling) != to.Bool(upstreamNodePool.EnableAutoScaling) {
+					logrus.Infof("Updating autoscaling in node pool [%s] for cluster [%s]", to.String(np.Name), spec.ClusterName)
+					needsUpdate = true
+				}
+				if np.OrchestratorVersion != nil && to.String(np.OrchestratorVersion) != to.String(upstreamNodePool.OrchestratorVersion) {
+					logrus.Infof("Updating orchestrator version in node pool [%s] for cluster [%s]", to.String(np.Name), spec.ClusterName)
+					needsUpdate = true
+				}
+			} else {
+				logrus.Infof("Adding node pool [%s] for cluster [%s]", to.String(np.Name), spec.ClusterName)
+				needsUpdate = true
 			}
-			if np.EnableAutoScaling != nil && *np.EnableAutoScaling {
-				agentProfile.EnableAutoScaling = np.EnableAutoScaling
-				agentProfile.MaxCount = np.MaxCount
-				agentProfile.MinCount = np.MinCount
+
+			if needsUpdate {
+				result, err := aks.CreateOrUpdateAgentPool(ctx, agentPoolClient, spec, &np)
+				if err != nil {
+					return spec, fmt.Errorf("failed to update cluster: %v", err)
+				}
+
+				if result.ManagedClusterAgentPoolProfileProperties.ProvisioningState == nil ||
+					*result.ManagedClusterAgentPoolProfileProperties.ProvisioningState == "Succeeded" {
+					logrus.Infof("Node pool [%s] in cluster [%s] was updated successfully", to.String(np.Name), spec.ClusterName)
+				}
+				needsUpdate = false
```

Comment:
nit - maybe use a different variable for the cluster update instead of resetting this variable? If the code gets moved around it would be easy to forget to make sure this reset is in the right place.

Reply:
I have changed it